presto-prometheus/src/main/java/io/prestosql/plugin/prometheus/PrometheusSplitManager.java

```diff
@@ -0,0 +1,285 @@
+                prometheusClient.config.getQueryChunkSizeDuration(),
+                table)
+                .forEach(time -> {
+                    try {
+                        splits.add(new PrometheusSplit(buildQuery(prometheusClient.config.getPrometheusURI(),
+                                time,
+                                table.getName(),
+                                prometheusClient.config.getQueryChunkSizeDuration())));
+                    }
+                    catch (URISyntaxException e) {
+                        throw new UnsupportedOperationException(e.getMessage());
+                    }
+                });
+        return new FixedSplitSource(splits);
+    }
+
+    // URIBuilder handles URI encode
+    private URI buildQuery(URI baseURI, String time, String metricName, String queryChunkSizeDuration)
+            throws URISyntaxException
+    {
+        List<NameValuePair> nameValuePairs = new ArrayList<>(2);
+        nameValuePairs.add(new BasicNameValuePair("query",
+                metricName + "[" + queryChunkSizeDuration + "]"));
+        nameValuePairs.add(new BasicNameValuePair("time",
+                time));
+        return new URIBuilder(baseURI.toString())
+                .setPath("api/v1/query")
+                .setParameters(nameValuePairs).build();
+    }
+
+    /**
+     * Utility method to get the end times in decimal seconds that divide up the query into chunks
+     * The times will be used in queries to Prometheus like: `http://localhost:9090/api/v1/query?query=up[21d]&time=1568229904.000"`
+     * ** NOTE: Prometheus instant query wants the duration and end time specified.
+     * We use now() for the defaultUpperBound when none is specified, for instance, from predicate pushdown
+     *
+     * @param defaultUpperBound         a LocalDateTime likely from PrometheusTimeMachine class for testability
+     * @param maxQueryRangeDurationStr  a String like `21d`
+     * @param queryChunkSizeDurationStr a String like `1d`
+     * @return list of end times as decimal epoch seconds, like ["1568053244.143", "1568926595.321"]
+     */
+    protected static List<String> generateTimesForSplits(LocalDateTime defaultUpperBound, String maxQueryRangeDurationStr, String queryChunkSizeDurationStr, PrometheusTable table)
+    {
+        Optional<PrometheusPredicateTimeInfo> predicateRange = table.getPredicate()
+                .flatMap(predicate -> determinePredicateTimes(predicate));
+
+        EffectiveLimits effectiveLimits = new EffectiveLimits(defaultUpperBound, maxQueryRangeDurationStr, predicateRange);
+        ZonedDateTime upperBound = effectiveLimits.getUpperBound();
+        java.time.Duration maxQueryRangeDuration = effectiveLimits.getMaxQueryRangeDuration();
+
+        java.time.Duration queryChunkSizeDuration = java.time.Duration.ofMillis(Duration.valueOf(queryChunkSizeDurationStr).toMillis());
+        if (maxQueryRangeDuration.isNegative()) {
```

Comment:
Validation should happen in config class

Reply:
There was doc string mistake on the param for this in the `EffectiveLimits` constructor that may have been obscuring that `maxQueryRangeDuration` is possibly from the config, but also possibly calculated in that constructor, so needs to be checked here. Doc `@param` value fixed in: b86bd28e